In [ ]:
import pandas as pd

df = pd.read_csv('/content/Coffe_sales.csv')
display(df.head())

hour_of_day cash_type  money    coffee_name Time_of_Day Weekday Month_name  \
0           10      card   38.7          Latte     Morning     Fri        Mar   
1           12      card   38.7  Hot Chocolate   Afternoon     Fri        Mar   
2           12      card   38.7  Hot Chocolate   Afternoon     Fri        Mar   
3           13      card   28.9      Americano   Afternoon     Fri        Mar   
4           13      card   38.7          Latte   Afternoon     Fri        Mar   

   Weekdaysort  Monthsort        Date             Time  
0            5          3  2024-03-01  10:15:50.520000  
1            5          3  2024-03-01  12:19:22.539000  
2            5          3  2024-03-01  12:20:18.089000  
3            5          3  2024-03-01  13:46:33.006000  
4            5          3  2024-03-01  13:48:14.626000

In [ ]:
display(df.info())
display(df.isnull().sum())
display(df.describe())
display(df.describe(include='object'))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3547 entries, 0 to 3546
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   hour_of_day  3547 non-null   int64  
 1   cash_type    3547 non-null   object 
 2   money        3547 non-null   float64
 3   coffee_name  3547 non-null   object 
 4   Time_of_Day  3547 non-null   object 
 5   Weekday      3547 non-null   object 
 6   Month_name   3547 non-null   object 
 7   Weekdaysort  3547 non-null   int64  
 8   Monthsort    3547 non-null   int64  
 9   Date         3547 non-null   object 
 10  Time         3547 non-null   object 
dtypes: float64(1), int64(3), object(7)
memory usage: 304.9+ KB


None

hour_of_day    0
cash_type      0
money          0
coffee_name    0
Time_of_Day    0
Weekday        0
Month_name     0
Weekdaysort    0
Monthsort      0
Date           0
Time           0
dtype: int64

hour_of_day        money  Weekdaysort    Monthsort
count  3547.000000  3547.000000  3547.000000  3547.000000
mean     14.185791    31.645216     3.845785     6.453905
std       4.234010     4.877754     1.971501     3.500754
min       6.000000    18.120000     1.000000     1.000000
25%      10.000000    27.920000     2.000000     3.000000
50%      14.000000    32.820000     4.000000     7.000000
75%      18.000000    35.760000     6.000000    10.000000
max      22.000000    38.700000     7.000000    12.000000

cash_type          coffee_name Time_of_Day Weekday Month_name  \
count       3547                 3547        3547    3547       3547   
unique         1                    8           3       7         12   
top         card  Americano with Milk   Afternoon     Tue        Mar   
freq        3547                  809        1205     572        494   

              Date             Time  
count         3547             3547  
unique         381             3547  
top     2024-10-11  18:11:38.635000  
freq            26                1

In [ ]:
# Drop the 'cash_type' column
df = df.drop('cash_type', axis=1)

# Convert 'Date' to datetime objects
df['Date'] = pd.to_datetime(df['Date'])

# Extract date features
df['day_of_week'] = df['Date'].dt.dayofweek
df['day_of_month'] = df['Date'].dt.day
df['month'] = df['Date'].dt.month

# Convert 'Time' to datetime objects and extract time features
df['Time'] = pd.to_datetime(df['Time']).dt.time
df['hour'] = df['Time'].apply(lambda t: t.hour)
df['minute'] = df['Time'].apply(lambda t: t.minute)


# Drop original date and time columns and related columns
df = df.drop(['Date', 'Time', 'Weekday', 'Month_name', 'Weekdaysort', 'Monthsort'], axis=1)

# Apply one-hot encoding
df = pd.get_dummies(df, columns=['coffee_name', 'Time_of_Day'], drop_first=True)

# Separate features and target variable
X = df.drop('money', axis=1)
y = df['money']

# Split data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

display(X_train.head())
display(X_test.head())
display(y_train.head())
display(y_test.head())

/tmp/ipython-input-357894816.py:13: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['Time'] = pd.to_datetime(df['Time']).dt.time


hour_of_day  day_of_week  day_of_month  month  hour  minute  \
1149           21            6            11      8    21      30   
195            14            4             5      4    14      23   
532            15            0            27      5    15      11   
2486           13            3            19     12    13      38   
2166           10            5             9     11    10       8   

      coffee_name_Americano with Milk  coffee_name_Cappuccino  \
1149                            False                   False   
195                             False                   False   
532                             False                   False   
2486                            False                   False   
2166                             True                   False   

      coffee_name_Cocoa  coffee_name_Cortado  coffee_name_Espresso  \
1149              False                 True                 False   
195               False                False                 False   
532               False                False                 False   
2486              False                False                 False   
2166              False                False                 False   

      coffee_name_Hot Chocolate  coffee_name_Latte  Time_of_Day_Morning  \
1149                      False              False                False   
195                       False               True                False   
532                       False              False                False   
2486                       True              False                False   
2166                      False              False                 True   

      Time_of_Day_Night  
1149               True  
195               False  
532               False  
2486              False  
2166              False

hour_of_day  day_of_week  day_of_month  month  hour  minute  \
299            19            1            23      4    19      43   
2550           21            2            25     12    21      57   
3079           17            2            19      2    17      34   
2342           14            5            30     11    14      16   
1211           15            4            16      8    15      18   

      coffee_name_Americano with Milk  coffee_name_Cappuccino  \
299                              True                   False   
2550                            False                   False   
3079                            False                    True   
2342                            False                   False   
1211                            False                   False   

      coffee_name_Cocoa  coffee_name_Cortado  coffee_name_Espresso  \
299               False                False                 False   
2550              False                False                 False   
3079              False                False                 False   
2342               True                False                 False   
1211              False                False                  True   

      coffee_name_Hot Chocolate  coffee_name_Latte  Time_of_Day_Morning  \
299                       False              False                False   
2550                      False               True                False   
3079                      False              False                False   
2342                      False              False                False   
1211                      False              False                False   

      Time_of_Day_Night  
299                True  
2550               True  
3079               True  
2342              False  
1211              False

1149    23.02
195     38.70
532     27.92
2486    35.76
2166    30.86
Name: money, dtype: float64

299     32.82
2550    35.76
3079    35.76
2342    35.76
1211    18.12
Name: money, dtype: float64

In [ ]:
from sklearn.ensemble import RandomForestRegressor

# Instantiate the model
model = RandomForestRegressor(n_estimators=100, random_state=42)

# Train the model
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate Mean Squared Error
mse = mean_squared_error(y_test, y_pred)

# Calculate R-squared score
r2 = r2_score(y_test, y_pred)

# Print the results
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared Score (R2): {r2:.2f}")

Mean Squared Error (MSE): 0.31
R-squared Score (R2): 0.99


## Summary:

### Data Analysis Key Findings

*   The dataset contains information on coffee sales, including details like cash type, coffee name, time of day, date, time, hour of day, weekday, month name, and the money received.
*   The initial data exploration revealed no missing values in the dataset.
*   The 'cash\_type' column had only one unique value ('card') and was subsequently dropped during preprocessing.
*   The most frequent coffee sold was 'Americano with Milk'.
*   After preprocessing, which included dropping irrelevant columns, extracting date and time features, and one-hot encoding categorical variables, the data was split into training and testing sets.
*   A RandomForestRegressor model was chosen and trained on the preprocessed training data.
*   The trained model achieved a Mean Squared Error (MSE) of 0.31 and an R-squared score (R2) of 0.99 on the test data, indicating a very good fit to the data.

### Insights or Next Steps

*   The high R-squared score suggests that the selected features and model are highly effective in predicting coffee sales (`money`).
*   Further analysis could explore feature importance to understand which features contribute most significantly to the predictions, potentially informing sales strategies.
